common/splice_script.c

```diff
@@ -0,0 +1,2696 @@
+			return new_error(ctx, INVALID_TOKEN, input[i],
+					 "operands");
+		}
+	}
+
+	tal_free(input);
+	tal_resize(&tokens, n);
+	*tokens_inout = tokens;
+	return NULL;
+}
+
+static struct splice_script_error *compress_2nd_operands(const tal_t *ctx,
+							 struct token ***tokens_inout)
+{
+	struct token **input = *tokens_inout;
+	struct token **tokens = tal_arr(ctx, struct token *, tal_count(input));
+	size_t n = 0;
+
+	for (size_t i = 0; i < tal_count(input); i++) {
+		switch(input[i]->type) {
+		case TOK_CHAR:
+		case TOK_ATSYM:
+		case TOK_COLON:
+			return new_error(ctx, INVALID_TOKEN, input[i],
+					 "2nd_operands");
+		case TOK_DELIMITER:
+		case TOK_ARROW:
+		case TOK_STR:
+		case TOK_SATS:
+		case TOK_PERCENT:
+		case TOK_QUESTION:
+		case TOK_WILDCARD:
+		case TOK_CHANID:
+		case TOK_WALLET:
+		case TOK_FEERATE:
+		case TOK_NODEID:
+		case TOK_BTCADDR:
+		case TOK_CHANQUERY:
+		case TOK_MULTI_CHANID:
+		case TOK_LEASERATE:
+			tokens[n++] = tal_steal(tokens, input[i]);
+			break;
+		case TOK_MINUS:
+		case TOK_PLUS:
+			if (i + 1 >= tal_count(input)
+				|| input[i+1]->type != TOK_FEE)
+				return new_error(ctx, MISSING_FEESTR, input[i],
+						 "2nd_operands");
+
+			/* This token is consumed. If negative, add flag to next
+			 * token */
+			if (input[i]->type == TOK_MINUS)
+				input[i+1]->flags |= TOKEN_FLAG_FEERATE_NEGATIVE;
+			break;
+		case TOK_FEE:
+			if (!n)
+				return new_error(ctx, INVALID_TOKEN, input[i],
+						 "2nd_operands");
+
+			/* We put FEE on the middle spot of the amount */
+			tokens[n-1]->middle = tal_steal(tokens[n-1], input[i]);
+			break;
+		case TOK_PIPE:
+			if (!n || i + 1 >= tal_count(input))
+				return new_error(ctx, INVALID_TOKEN, input[i],
+						 "2nd_operands");
+			input[i]->type = TOK_LEASEREQ;
+			input[i]->right = tal_steal(input[i], input[i+1]);
+
+			/* We put LEASEREQ on the right spot of the amount */
+			tokens[n-1]->right = tal_steal(tokens[n-1], input[i]);
+			i++;
+			break;
+		case TOK_LEASEREQ:
+		case TOK_SEGMENT:
+			return new_error(ctx, INVALID_TOKEN, input[i],
+					 "2nd_operands");
+		}
+	}
+
+	tal_free(input);
+	tal_resize(&tokens, n);
+	*tokens_inout = tokens;
+	return NULL;
+}
+
+static bool matches_chan_id(struct token *token, struct channel_id chan_id)
+{
+	if (token->chan_id && channel_id_eq(token->chan_id, &chan_id))
+		return true;
+
+	if (token->left && matches_chan_id(token->left, chan_id))
+		return true;
+
+	if (token->middle && matches_chan_id(token->middle, chan_id))
+		return true;
+
+	if (token->right && matches_chan_id(token->right, chan_id))
+		return true;
+
+	return false;
+}
+
+/* Searches through both tokensA and tokensB. */
+static struct node_id *first_node_with_unused_chan(const tal_t *ctx,
+						   struct splice_script_chan *channels,
+						   size_t channels_count,
+						   struct token **tokensA,
+						   size_t a_size,
+						   struct token **tokensB,
+						   size_t b_size)
+{
+	for (size_t i = 0; i < channels_count; i++) {
```

Comment:
sort before going thru it?? global sort on stash?

Reply:
todo: make a new method on the channel fetcher (wallet.c) that always returns sorted channels by oldest to newest.

Reply:
Duplicate of https://github.com/ElementsProject/lightning/pull/6980#discussion_r1621356185